[WIP] exp save: initial implementation

dtrifiro opened the pull request:
Makes it possible to save changes in the current workspace as a dvc experiment.

closes #6746

dberenbaum commented:
Looks good so far! I think `--metrics` is okay to skip for now if it's easier and `--json` isn't needed.

I ran into a couple issues:

1. How to overwrite experiments. `dvc exp save -f` should have the same effect as `dvc exp run -f` here:

```
$ dvc exp save -f -n small_batch
ERROR: failed to save experiment - Experiment conflicts with existing experiment 'small_batch'. To overwrite the existing experiment run:

        dvc exp run -f ...

To run this experiment with a different name run:

        dvc exp run -n <new_name> ...

```


2. When saving an experiment with files that weren't yet tracked by Git, they aren't showing up as expected. Here's an example on a fresh repo:

```
$ dvc stage add -q -n exp_save -M metrics.yaml 'echo "foo: 1" > metrics.yaml'
$ dvc repro -q
$ dvc exp save -q
$ dvc exp show
 ──────────────────────────────────────────
  Experiment                Created    foo
 ──────────────────────────────────────────
  workspace                 -            1
  main                      11:59 AM     -
  └── 74a5ea2 [exp-48145]   11:59 AM     -
 ──────────────────────────────────────────
$ git add .
$ git commit -qm "add exp to git"
$ dvc exp save -f -q
$ dvc exp show
 ──────────────────────────────────────────
  Experiment                Created    foo
 ──────────────────────────────────────────
  workspace                 -            1
  main                      12:00 PM     1
  └── e51764f [exp-48145]   12:00 PM     1
 ──────────────────────────────────────────
```

dtrifiro commented:
Thanks @dberenbaum. I fixed the first issue.

Regarding the second. It looks to me like it's the expected behaviour. Here's a similar example with `exp run` (on [example-get-started](https://github.com/iterative/example-get-started))

```console
→ dvc exp run -S train.seed=42 --name on-main
...
Ran experiment(s): on-main
Experiment results have been applied to your workspace.

To promote an experiment to a Git branch run:

        dvc exp branch <exp> <branch>

→ dvc exp list
main:
        on-main
→ touch antani && git add antani && git commit -m "antani" 
[main 00d99d3] antani
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 antani
→ dvc exp list # this show no exps for the current baseline
→ dvc exp run -S train.seed=42 --name on-updated-main
...
Ran experiment(s): on-updated-main
Experiment results have been applied to your workspace.

To promote an experiment to a Git branch run:

        dvc exp branch <exp> <branch>

→ dvc exp list # we see the new experiments
main:
        on-updated-main
→ dvc exp list -A # running with --all will show experiments for all baselines
9914c47:
        on-main
main:
        on-updated-main
```

Also added a test (`test_exp_save_after_commit`) if you want to have a look

daavoo commented on dvc/repo/experiments/executor/local.py:
```diff
@@ -213,3 +215,58 @@ def cleanup(self, infofile: str):
+        exp_hash: Optional[str] = None
+        exp_ref: Optional[ExpRefInfo] = None
+
+        dvc = Repo(os.path.join(info.root_dir, info.dvc_dir))
```
Why create a `Repo` instead of receiving as arg?

pmrowla replied:
Executors are set up to do it this way (without ever passing complex classes like `Repo`) so that they work easily in multiprocessing environments. We can't pass `Repo` because it's not pickle-able.

daavoo commented on dvc/repo/experiments/executor/local.py:
```diff
@@ -213,3 +215,57 @@ def cleanup(self, infofile: str):
+            ref: Optional[str] = dvc.scm.get_ref(EXEC_BRANCH, follow=False)
+            exp_ref = ExpRefInfo.from_ref(ref) if ref else None
+            # TODO: research into how untracked files should be handled
+            if cls.WARN_UNTRACKED:
```
> Add a flag to save that forces untracked files to be added included in the experiment

I am not sure what makes more sense from product perspective but I would expect untracked files to be included by default? c @dberenbaum

pmrowla replied:
We exclude untracked files from regular experiments, I would not expect `exp save` to include untracked files.

(including untracked files generally leads to unwanted behavior like git adding things like `__pycache__` directories that users forgot to gitignore)

daavoo replied:
> We exclude untracked files from regular experiments, I would not expect exp save to include untracked files.

I was thinking that saving untracked in `exp save` could make sense because the use case is different. Some requirements in the original issue also sounded like referring to untracked files, i.e.:

> Code has already been run. The user might have already run code and wants to record it without committing it to their working branch.

dberenbaum replied:
Whether exp saves or doesn't save untracked files, I think it should be consistent between `exp run` and `exp save`, and we need to make it very clear to users what the behavior is. Related: https://github.com/iterative/dvc.org/issues/4044.

daavoo replied:
In order to use `exp.save` from a Python API (DVCLive scenario), it would be great if untracked files could be included (or an option exposed to force it). 

Otherwise, the first time DVCLive is run (or something new is logged that was not tracked before) the logged data will not be present in the exp ref:

```
from dvclive import Live

live = Live()
live.log_metric("foo", 1)

# This should call something like `dvc.api.make_experiment()`
# and the untracked files logged by DVCLive should be included
live.end()
```

dberenbaum replied:
I think we can split between:
1. DVCLive-generated files: can DVCLive handle tracking these with Git?
2. Other untracked files: it's possible model artifacts or other files are being saved outside DVCLive, so it would be useful to at least have a flag to include these. I'd be wary of including everything in the repo by default.

Let's make 2 a follow up and worry about 1 for now.

daavoo replied:
> DVCLive-generated files: can DVCLive handle tracking these with Git?

I guess so. It just feels odd.

> Let's make 2 a follow up and worry about 1 for now.

Ok, will implement 1 on DVCLive side

daavoo commented on dvc/repo/experiments/save.py:
```diff
@@ -0,0 +1,32 @@
+    queue = repo.experiments.workspace_queue
+    logger.debug("Saving workspace in %s", os.getcwd())
+
+    entry = repo.experiments.new(queue=queue, name=name, force=force)
+    executor = queue.init_executor(repo.experiments, entry)
+    save_result = executor.save(executor.info, force=force)
+    result = queue.collect_executor(repo.experiments, executor, save_result)
```
Feels a little odd to involve stuff like executor and queue when we are not really going to execute or queue anything. 

I get that is because `_stash_exp` is currently part of the `queue` implementation and it might be out of the scope of the P.R. to refactor, but wanted to discuss it.

pmrowla replied:
The workspace queue is already a dummy queue that only ever contains one item.

I think this actually does work well since using the full executor init/collect workflow this way sets up required git refs for generating exp commits and refs (and cleans up the git workspace when we are done). Doing it this way also ensures that the workspace state is restored in the event that something in DVC breaks during the actual save commit operation(s).

dberenbaum commented:
Another issue I'm seeing is that the merge fails if I actually stage my Git changes with `git add`:

```
$ dvc stage add -q -n exp_save -M metrics.yaml 'echo "foo: 1" > metrics.yaml'

$ dvc repro
Running stage 'exp_save':
> echo "foo: 1" > metrics.yaml
Generating lock file 'dvc.lock'
Updating lock file 'dvc.lock'

To track the changes with git, run:

        git add dvc.lock

To enable auto staging, run:

        dvc config core.autostage true
Use `dvc push` to send your updates to remote storage.

$ git add .

$ dvc exp save -v
2022-10-17 11:29:02,737 DEBUG: Saving workspace in /Users/dave/repo
2022-10-17 11:29:02,801 DEBUG: Stashed experiment 'dd42a95' with baseline '9f7aa79' for future execution.
2022-10-17 11:29:02,806 DEBUG: Init workspace executor in '/Users/dave/repo'
2022-10-17 11:29:02,810 ERROR: failed to save experiment - Merge failed
------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dave/miniforge3/envs/tensorflow/lib/python3.10/site-packages/scmrepo/git/backend/pygit2.py", line 694, in merge
    self.repo.merge(obj.id)
_pygit2.GitError: 3 uncommitted changes would be overwritten by merge

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/dave/Code/dvc/dvc/repo/experiments/executor/local.py", line 196, in init_git
    self.scm.merge(merge_rev, squash=True, commit=False)
  File "/Users/dave/miniforge3/envs/tensorflow/lib/python3.10/site-packages/scmrepo/git/__init__.py", line 289, in _backend_func
    result = func(*args, **kwargs)
  File "/Users/dave/miniforge3/envs/tensorflow/lib/python3.10/site-packages/scmrepo/git/backend/pygit2.py", line 697, in merge
    raise SCMError("Merge failed") from exc
scmrepo.exceptions.SCMError: Merge failed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/dave/Code/dvc/dvc/commands/experiments/save.py", line 16, in run
    ref = self.repo.experiments.save(
  File "/Users/dave/Code/dvc/dvc/repo/experiments/__init__.py", line 528, in save
    return save(self.repo, *args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/save.py", line 27, in save
    executor = queue.init_executor(repo.experiments, entry)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/utils.py", line 40, in wrapper
    return f(exp, *args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/repo/experiments/queue/base.py", line 584, in init_executor
    executor.init_git(
  File "/Users/dave/Code/dvc/dvc/repo/experiments/executor/local.py", line 198, in init_git
    raise GitMergeError(str(exc), scm=self.scm)
dvc.scm.GitMergeError: Merge failed
------------------------------------------------------------
2022-10-17 11:29:02,815 DEBUG: Analytics is disabled.
```

daavoo commented:
One of the use cases for this is in DVCLive onboarding, where the only requisite would be `dvc init` (`dvc.yaml` is not required).

In that scenario:

- DVC might be installed as a binary, so internal Python API won't be available unless the user also installs the `dvc` Python lib.
- We can consider that there are not DVC tracked artifacts or pipeline, so DVC operations (i.e. `dvc commit`) are not needed.

It would be great to consider how much of the exp save logic could be isolated as pure `scmrepo` operations, as it might be convenient for DVLive to only have `scmrepo` as a requirement instead of full DVC.

dberenbaum commented:
> It would be great to consider how much of the exp save logic could be isolated as pure `scmrepo` operations, as it might be convenient for DVLive to only have `scmrepo` as a requirement instead of full DVC.

It's a good idea, but we should probably make it a follow-up issue and suggest users install as a Python library for now.

daavoo commented on dvc/repo/experiments/executor/local.py:
```diff
@@ -213,3 +215,58 @@ def cleanup(self, infofile: str):
+
+        try:
+            stages = dvc.commit([], force=force)
+            exp_hash = cls.hash_exp(stages)
```
Another comment from the DVCLive scenario (we can discuss if we want to address those in a follow-up) cc @dberenbaum 

The current way of commuting exp_hash is problematic in the "no pipeline" because there are no stages in the `dvc.yaml` (or even there is no `dvc.yaml`), so multiple calls to `experiments.save` will result in a conflict (unless a custom name is passed).

In a repo with no `dvc.yaml`, the following fails with `dvc.repo.experiments.exceptions.ExperimentExistsError`:

```py
from dvc.repo import Repo

repo = Repo()

for i in range(3):
    with open("params.yaml", "w") as f:
        f.write(f"foo: {i}")
    repo.experiments.save()
```

But this works as expected:


```py
from dvc.repo import Repo

repo = Repo()

for i in range(3):
    with open("params.yaml", "w") as f:
        f.write(f"foo: {i}")
    repo.experiments.save(name=str(i))
```

```console
$ dvc exp show
  Experiment        Created    foo  
 ────────────────────────────────── 
  workspace         -          2    
  master            10:28 AM   2    
  ├── 71d48a5 [2]   10:37 AM   2    
  ├── f87cc9e [1]   10:37 AM   1    
  └── 7adc787 [0]   10:37 AM   0    
 ────────────────────────────────── 
```

dberenbaum replied:
I assume it's more about `dvc.lock` than `dvc.yaml`?

I guess we could have DVCLive auto-generate names for now (related to https://github.com/iterative/vscode-dvc/issues/2457) if it's easier than revisiting how exp names are assigned, but it might be better to track changes across the entire repo instead of only in `dvc.lock`.

daavoo commented:
Created a new P.R. https://github.com/iterative/dvc/pull/8599 because this was created from @dtrifiro repo so I could not force push. The new P.R. is a branch in the repo itself instead of coming from a fork

dberenbaum commented:
> > Another issue I'm seeing is that the merge fails if I actually stage my Git changes with git add:
> 
> What behavior would you expect here?
> 
> Something similar to `exp run`, where a warning is shown and the changes are unstaged?:

🤔 I can't remember why `dvc exp run` unstages changes that the user staged. It seems more intuitive to keep whatever was staged in the same state, but that's not my main concern here.

More importantly, `exp save` should succeed, regardless of whether those files remain staged afterwards.
